src/Services/Dominion/Actions/EspionageActionService.php

```diff
@@ -386,6 +409,204 @@ protected function performInfoGatheringOperation(Dominion $dominion, string $ope
         ];
     }
 
-    // todo: theft/black ops/war
+    protected function performResourceTheftOperation(Dominion $dominion, string $operationKey, Dominion $target): array
+    {
+        $operationInfo = $this->espionageHelper->getOperationInfo($operationKey);
+
+        $selfSpa = $this->militaryCalculator->getSpyRatio($dominion, 'offense');
+        $targetSpa = $this->militaryCalculator->getSpyRatio($target, 'defense');
+
+        // You need at least some positive SPA to perform espionage operations
+        if ($selfSpa === 0.0) {
+            return [
+                'success' => false,
+                'message' => "Your spy force is too weak to cast {$operationInfo['name']}. Please train some more spies.",
+                'alert-type' => 'warning',
+            ];
+        }
+
+        if ($targetSpa !== 0.0) {
+            $ratio = ($selfSpa / $targetSpa);
+
+            // todo: copied from spell success ratio. needs looking into later
+            // todo: factor in spy strength
+            $successRate = clamp((
+                (0.0172 * ($ratio ** 3))
+                - (0.1809 * ($ratio ** 2))
+                + (0.6767 * $ratio)
+                - 0.0134
+            ), 0.0, 1.0);
+
+            if (!random_chance($successRate)) {
+                // todo: move to CasualtiesCalculator
+
+                // Values (percentage)
+                $spiesKilledBasePercentage = 0.2; // TODO: Higher for black ops.
```

Comment:
2% spies killed on failed theft op? Just checking

Reply:
This is just copied from info ops, probably should be increased? I don't see anything about the number killed in the Wiki etc.

Reply:
```suggestion
                $spiesKilledBasePercentage = 1;
```

Discussed on Discord with BR. Gonna go with the following for now:

- 0.2% failed spy ops
- 1% failed resource theft ops <--
- 2% failed black ops